File: Difficulty: Medium/Aggressive Cows/aggressive-cows.cpp
class Solution {
  public:
    bool canput(vector<int> &arr, int k,int diff){
        
        int curr=1 ; int last = arr[0];
        
        for(int i = 1 ; i<arr.size();i++){
            
            if(arr[i]-last>=diff){
                curr++;
                last = arr[i];
            }
        }
        
        if(curr>=k)return true;
        return false;
    }
    int aggressiveCows(vector<int> &arr, int k) {
        // code here
        
        int n = arr.size();
        sort(arr.begin(),arr.end());
        int low = 1;
        int high = arr[n-1]-arr[0];
        
        int ans = INT_MIN;
        
        
        while(low<=high){
            
            int mid = low + (high-low)/2;
            
            if(canput(arr,k,mid)){
                ans=max(ans,mid);
                low=mid+1;
            }else{
                high=mid-1;
            }
        }
        
        return ans;
    }
};

File: Difficulty: Medium/Aggressive Cows/aggressive-cows.py
class Solution:
    def canput(self,arr,k,mid):
        
        count =1
        last = arr[0]
        
        for i in range(1,len(arr)):
            if arr[i]-last>=mid:
                count+=1
                last=arr[i]
        
        if count<k : return False
        return True
    def aggressiveCows(self, arr, k):
        # code here
        arr.sort()
        n=len(arr)
        low = 1
        high=arr[n-1]
        ans=float("-inf")
        
        
        while low<=high:
            
            mid = low + (high-low)//2
            if(self.canput(arr,k,mid)):
                ans=max(ans,mid)
                low=mid+1
            else:
                high=mid-1
        return ans
                
        
        